File: uniclub-api/schemas.py
from pydantic import BaseModel, Field, EmailStr, ConfigDict
from typing import Optional, List
from datetime import date, datetime
from models import EventStatus, BoardRole


class ResponseSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

# ==========================
# CLUB SCHEMAS
# ==========================
class ClubCreate(BaseModel):
    name: str = Field(..., min_length=1)
    description: str = Field(..., min_length=1)
    category: str
    founded_date: date

class ClubResponse(ResponseSchema):
    id: int
    name: str
    description: str
    category: str
    founded_date: date

class ClubUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1)
    description: Optional[str] = Field(None, min_length=1)
    category: Optional[str] = None
    founded_date: Optional[date] = None

# ==========================
# ADVISOR SCHEMAS
# ==========================
class AdvisorCreate(BaseModel):
    full_name: str = Field(..., min_length=1)
    email: EmailStr
    department: str = Field(..., min_length=1)
    assigned_date: date
    club_id: Optional[int] = None

class AdvisorResponse(ResponseSchema):
    id: int
    full_name: str
    email: EmailStr
    department: str
    assigned_date: date
    club_id: Optional[int] = None

# ==========================
# MEMBER SCHEMAS
# ==========================
class MemberCreate(BaseModel):
    student_id: str = Field(..., min_length=1, pattern=r"^\w+$")
    first_name: str = Field(..., min_length=1)
    last_name: str = Field(..., min_length=1)
    email: EmailStr
    department: str = Field(..., min_length=1)
    join_date: date
    club_id: Optional[int] = None

class MemberResponse(ResponseSchema):
    id: int
    student_id: str
    first_name: str
    last_name: str
    email: EmailStr
    department: str
    join_date: date
    leave_date: Optional[date] = None
    club_id: Optional[int] = None

class MemberUpdateLeaveDate(BaseModel):
    leave_date: date

# ==========================
# BOARD MEMBER SCHEMAS
# ==========================
class BoardMemberCreate(BaseModel):
    student_id: str = Field(..., min_length=1)
    first_name: str = Field(..., min_length=1)
    last_name: str = Field(..., min_length=1)
    email: EmailStr
    role: BoardRole
    join_date: date
    club_id: int

class BoardMemberResponse(ResponseSchema):
    id: int
    student_id: str
    first_name: str
    last_name: str
    email: EmailStr
    role: BoardRole
    join_date: date
    leave_date: Optional[date] = None
    club_id: int

# ==========================
# VENUE SCHEMAS
# ==========================
class VenueCreate(BaseModel):
    name: str = Field(..., min_length=1)
    location: str = Field(..., min_length=1)
    capacity: int = Field(..., gt=0)
    venue_type: Optional[str] = None
    description: Optional[str] = None

class VenueResponse(ResponseSchema):
    id: int
    name: str
    location: str
    capacity: int
    venue_type: Optional[str] = None
    description: Optional[str] = None

# ==========================
# EVENT SCHEMAS
# ==========================
class EventCreate(BaseModel):
    title: str = Field(..., min_length=1)
    description: str = Field(..., min_length=1)
    status: EventStatus
    event_start: datetime
    event_end: datetime
    club_id: int
    venue_id: Optional[int] = None

class EventUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1)
    description: Optional[str] = Field(None, min_length=1)
    status: Optional[EventStatus] = None
    event_start: Optional[datetime] = None
    event_end: Optional[datetime] = None
    club_id: Optional[int] = None
    venue_id: Optional[int] = None

class EventResponse(ResponseSchema):
    id: int
    title: str
    description: str
    status: EventStatus
    event_start: datetime
    event_end: datetime
    club_id: int
    venue_id: Optional[int] = None

# ==========================
# BUDGET SCHEMAS
# ==========================
class BudgetCreate(BaseModel):
    event_id: int
    planned_amount: float = Field(..., ge=0)
    actual_amount: float = Field(..., ge=0)
    notes: Optional[str] = None

class BudgetResponse(ResponseSchema):
    id: int
    event_id: int
    planned_amount: float
    actual_amount: float
    notes: Optional[str] = None

class BudgetUpdate(BaseModel):
    planned_amount: Optional[float] = Field(None, ge=0)
    actual_amount: Optional[float] = Field(None, ge=0)
    notes: Optional[str] = None

# ==========================
# REGISTRATION SCHEMAS
# ==========================
class RegistrationCreate(BaseModel):
    event_id: int
    member_id: int
    registered_at: Optional[datetime] = None

class RegistrationResponse(ResponseSchema):
    id: int
    event_id: int
    member_id: int
    registered_at: datetime

# ==========================
# PARTICIPANT SCHEMAS
# ==========================
class ParticipantCreate(BaseModel):
    first_name: str = Field(..., min_length=1)
    last_name: str = Field(..., min_length=1)
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    event_id: int
    member_id: Optional[int] = None
    checked_in_at: Optional[datetime] = None

class ParticipantResponse(ResponseSchema):
    id: int
    first_name: str
    last_name: str
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    event_id: int
    member_id: Optional[int] = None
    checked_in_at: Optional[datetime] = None

# ==========================
# MESSAGE SCHEMAS
# ==========================
class MessageCreate(BaseModel):
    subject: str = Field(..., min_length=1)
    content: str = Field(..., min_length=1)
    club_id: int
    member_id: int
    sent_at: Optional[datetime] = None

class MessageResponse(ResponseSchema):
    id: int
    subject: str
    content: str
    club_id: int
    member_id: int
    sent_at: datetime

# ==========================
# SPONSORSHIP SCHEMAS
# ==========================
class SponsorshipCreate(BaseModel):
    sponsor_name: str = Field(..., min_length=1)
    amount: float = Field(..., ge=0)
    agreement_date: date
    event_id: int

class SponsorshipResponse(ResponseSchema):
    id: int
    sponsor_name: str
    amount: float
    agreement_date: date
    event_id: int


# ==========================
# REPORT SCHEMAS
# ==========================
class ClubNetworkReport(ResponseSchema):
    club: ClubResponse
    advisor: Optional[AdvisorResponse] = None
    members: List[MemberResponse]
    board_members: List[BoardMemberResponse]
    events: List[EventResponse]
    messages: List[MessageResponse]
    counts: dict[str, int]


class EventNetworkReport(ResponseSchema):
    event: EventResponse
    venue: Optional[VenueResponse] = None
    budget: Optional[BudgetResponse] = None
    registrations: List[RegistrationResponse]
    participants: List[ParticipantResponse]
    sponsorships: List[SponsorshipResponse]
    counts: dict[str, int]


class MemberNetworkReport(ResponseSchema):
    member: MemberResponse
    club: Optional[ClubResponse] = None
    messages: List[MessageResponse]
    registrations: List[RegistrationResponse]
    participant_records: List[ParticipantResponse]
    counts: dict[str, int]

File: uniclub-web/src/validation/schemas.ts
import { z } from "zod";

// ── Club ──
export const clubSchema = z.object({
  name: z.string().min(1, "Name is required"),
  description: z.string().min(1, "Description is required"),
  category: z.string().min(1, "Category is required"),
  founded_date: z.string().min(1, "Founded date is required"),
});
export type ClubFormValues = z.infer<typeof clubSchema>;

// ── Advisor ──
export const advisorSchema = z.object({
  full_name: z.string().min(1, "Full name is required"),
  email: z.string().email("Valid email is required"),
  department: z.string().min(1, "Department is required"),
  assigned_date: z.string().min(1, "Assigned date is required"),
  club_id: z
    .number()
    .int()
    .positive()
    .optional(),
});
export type AdvisorFormValues = z.infer<typeof advisorSchema>;

// ── Member ──
export const memberSchema = z.object({
  student_id: z
    .string()
    .min(1, "Student ID is required")
    .regex(/^\w+$/, "Student ID must be alphanumeric"),
  first_name: z.string().min(1, "First name is required"),
  last_name: z.string().min(1, "Last name is required"),
  email: z.string().email("Valid email is required"),
  department: z.string().min(1, "Department is required"),
  join_date: z.string().min(1, "Join date is required"),
  club_id: z
    .number()
    .int()
    .positive()
    .optional(),
});
export type MemberFormValues = z.infer<typeof memberSchema>;

// ── Board Member ──
export const boardMemberSchema = z.object({
  student_id: z.string().min(1, "Student ID is required"),
  first_name: z.string().min(1, "First name is required"),
  last_name: z.string().min(1, "Last name is required"),
  email: z.string().email("Valid email is required"),
  role: z.enum(
    ["President", "Vice President", "Secretary", "Treasurer", "Coordinator"],
    { message: "Role is required" },
  ),
  join_date: z.string().min(1, "Join date is required"),
  club_id: z.number().int().positive("Club is required"),
});
export type BoardMemberFormValues = z.infer<typeof boardMemberSchema>;

// ── Venue ──
export const venueSchema = z.object({
  name: z.string().min(1, "Name is required"),
  location: z.string().min(1, "Location is required"),
  capacity: z.number().int().positive("Capacity must be greater than 0"),
  venue_type: z.string().optional(),
  description: z.string().optional(),
});
export type VenueFormValues = z.infer<typeof venueSchema>;

// ── Event ──
export const eventSchema = z
  .object({
    title: z.string().min(1, "Title is required"),
    description: z.string().min(1, "Description is required"),
    status: z.enum(["Scheduled", "Completed", "Canceled"]),
    event_start: z.string().min(1, "Event start is required"),
    event_end: z.string().min(1, "Event end is required"),
    club_id: z.number().int().positive(),
    venue_id: z.number().int().positive().optional(),
  })
  .superRefine((value, ctx) => {
    if (new Date(value.event_end) <= new Date(value.event_start)) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["event_end"],
        message: "Event end must be after event start",
      });
    }
  });
export type EventFormValues = z.infer<typeof eventSchema>;

// ── Budget ──
export const budgetSchema = z.object({
  event_id: z.number().int().positive("Event is required"),
  planned_amount: z.number().min(0, "Planned amount must be 0 or greater"),
  actual_amount: z.number().min(0, "Actual amount must be 0 or greater"),
  notes: z.string().optional(),
});
export type BudgetFormValues = z.infer<typeof budgetSchema>;

// ── Registration ──
export const registrationSchema = z.object({
  event_id: z.number().int().positive("Event is required"),
  member_id: z.number().int().positive("Member is required"),
});
export type RegistrationFormValues = z.infer<typeof registrationSchema>;

// ── Message ──
export const messageSchema = z.object({
  subject: z.string().min(1, "Subject is required"),
  content: z.string().min(1, "Content is required"),
  club_id: z.number().int().positive("Club is required"),
  receiver_user_id: z.number().int().positive("Recipient is required"),
});
export type MessageFormValues = z.infer<typeof messageSchema>;

// ── Sponsorship ──
export const sponsorshipSchema = z.object({
  sponsor_name: z.string().min(1, "Sponsor name is required"),
  amount: z.number().min(0, "Amount must be 0 or greater"),
  agreement_date: z.string().min(1, "Agreement date is required"),
  event_id: z.number().int().positive("Event is required"),
});
export type SponsorshipFormValues = z.infer<typeof sponsorshipSchema>;

// ── Participant ──
export const participantSchema = z.object({
  first_name: z.string().min(1, "First name is required"),
  last_name: z.string().min(1, "Last name is required"),
  email: z
    .string()
    .email("Invalid email")
    .optional()
    .or(z.literal("")),
  phone: z.string().optional(),
  event_id: z.number().int().positive("Event is required"),
  member_id: z
    .number()
    .int()
    .positive()
    .optional(),
});
export type ParticipantFormValues = z.infer<typeof participantSchema>;
